### PR TITLE
Add a node for selecting specific shells from preprocessed dwis

### PR DESCRIPTION
**Pull request by mattcieslak**
## Background

The SS3T workflow needs to have only b=0s and one shell of b>0s sent to it. Similarly, the BabyAFQ pipeline requires selecting only a few shells for msmt.

## Recon Spec

This PR adds a `select_gradients` action that lets users pick which shells to send to a downstream node. The parameters are 

  1. A list of shells to select. These can be specific numeric b-values or the strings "highest" for highest b-value shell or "lowest" for lowest b>0 shell. If no b=0 shell is selected then one is added in the workflow.
  2. A parameter for how different the b-values can be from the b-values selected in (1). There are usually some differences in the acquired b-value and the requested b-value. This parameter is similar to the --b0-threshold parameter from qsiprep.
  3. If the user wants to do a sanity check on their input data, they can also specify the `expected_n_input_shells` to make sure that this many shells (including the b=0s as a shell) are detected in the input data.

An example in a recon spec for HBCD might look like:

```yaml
-   action: select_gradients
    input: qsirecon
    name: select_single_shell
    parameters:
        requested_shells:
            - 0
            - highest
        bval_distance_cutoff: 100
        expected_n_input_shells: 5
```

## Implementation

The trickiest part about this is making it general enough that it can be included in a built-in recon workflow and work with the many different shell schemes out there. The value `0` can be hard-coded because has a specific meaning/use in DWI. Because of the minor variations in b-values we also don't always know immediately what the intended shell b-values are. So the steps for selecting shells are 

  1. Figure out how many shells there are using the bval_distance_cutoff and agglomerative clustering with complete linkage
  2. Select the median b-value per cluster as the intended b-value
  3. loop over the requested bvals and find which shells they correspond to

## Tests and reports

The implementation in this PR creates a Pandas DataFrame with each b-value in the input data and which shell it corresponds to. I think saving this data frame to a tsv would be a sufficient sanity check for users. We had talked about plots with histograms of b-values per assigned shell, but these end up not looking very good/informative when plotted.

### CODE REVIEW

File: qsirecon/interfaces/gradients.py [hunk 1]
```diff
@@ -18,33 +18,201 @@
...
 
 
-class RemoveDuplicatesInputSpec(BaseInterfaceInputSpec):
+class _GradientSelectInputSpec(BaseInterfaceInputSpec):
     dwi_file = File(exists=True, mandatory=True)
     bval_file = File(exists=True, mandatory=True)
     bvec_file = File(exists=True, mandatory=True)
+    b_file = File(exists=True, mandatory=True)
+    btable_file = File(exists=True, mandatory=True)
+    bval_distance_cutoff = traits.Float(5.0, usedefault=True)
```
Comment (tsalo) [inline qsirecon/interfaces/gradients.py:27]:
```suggestion
    bval_distance_cutoff = traits.Float(
        5.0,
        usedefault=True,
        desc='Distance from the center of each b-value cluster to consider within a given shell',
    )
```

File: qsirecon/interfaces/gradients.py [hunk 2]
```diff
@@ -18,33 +18,201 @@
...
+        linkage="complete",
+    ).fit(bvals.reshape(-1, 1))
+
+    bval_df = pd.DataFrame({"bvalue": bvals, "assignment": agg_cluster.labels_})
+    shell_df = bval_df.groupby("assignment", as_index=False).agg({"bvalue": "median"})
+    bval_df["assigned_shell"] = bval_df["assignment"].replace(
+        shell_df["assignment"].tolist(), shell_df["bvalue"].tolist()
+    )
+    bval_df["shell_num"] = bval_df["assigned_shell"].rank(method="dense")
+    bval_df.drop("assigned_shell", inplace=True)
```
Comment (tsalo) [inline qsirecon/interfaces/gradients.py:182]:
```suggestion
    bval_df.drop(columns=["assigned_shell"], inplace=True)
```
I had to make this change to get it working locally.

File: qsirecon/interfaces/gradients.py [hunk 3]
```diff
@@ -18,33 +18,201 @@
...
+    import pandas as pd
+    from sklearn.cluster import AgglomerativeClustering
+
+    agg_cluster = AgglomerativeClustering(
+        n_clusters=None,
+        distance_threshold=2 * max_distance,
+        linkage="complete",
+    ).fit(bvals.reshape(-1, 1))
+
+    bval_df = pd.DataFrame({"bvalue": bvals, "assignment": agg_cluster.labels_})
```
Comment (tsalo) [inline qsirecon/interfaces/gradients.py:176]:
```suggestion
    from sklearn.cluster import AgglomerativeClustering
    from sklearn.metrics import silhouette_score

    X = bvals.reshape(-1, 1)
    agg_cluster = AgglomerativeClustering(
        n_clusters=None,
        distance_threshold=2 * max_distance,
        linkage="complete",
    ).fit(X)
    shells = agg_cluster.labels_
    
    score = silhouette_score(X, shells)
    if score < 0.8:
        print('Silhouette score is low. Is this is a DSI scheme?')

    bval_df = pd.DataFrame({"bvalue": bvals, "assignment": shells})
```

**Review by tsalo — COMMENTED** [review]
A few initial thoughts.

The big thing I think this needs is a comprehensive unit test for `_find_shells` with a bunch of simulated b-val schemes.

File: qsirecon/interfaces/gradients.py (hunk 1 above)
Comment (mattcieslak) [reply qsirecon/interfaces/gradients.py:27]: This is a different interface, which I don't believe is ever actually used in recon workflows. This particular distance cutoff is to determine whether two or more images are sampling approximately the same coordinate in q space.

File: qsirecon/tests/test_interfaces.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import nibabel as nb
+import numpy as np
+import pytest
+
+from qsirecon.interfaces.gradients import GradientSelect
+from qsirecon.tests.utils import download_test_data
+
+
+@pytest.mark.integration
```
Comment (tsalo) [inline qsirecon/tests/test_interfaces.py:13]: I wouldn't call this an integration test. I normally reserve that for the tip-to-tail workflow tests.

File: qsirecon/interfaces/gradients.py [hunk 5]
```diff
@@ -18,33 +18,214 @@
...
+                    bval_df["assigned_shell"] <= requested_bval + max_distance,
+                )
+            ]
+            if len(ok_shells.unique()) > 1:
+                raise Exception(
+                    f"Unable to unambiguously select b={requested_bval}."
+                    f"instead select from {bval_df.assigned_shells.unique().tolist()}"
+                )
+
+            numeric_bvals.append(ok_shells.iloc[0])
```
Comment (tsalo) [inline qsirecon/interfaces/gradients.py:152]: Shouldn't this check apply across requested bval options?

File: qsirecon/tests/test_interfaces.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    assert abcd_df["shell_num"].max() == 5.0
+
+    # HCP has 3 shells + b=0s
+    hcp_bval = np.loadtxt(load_data("schemes/HCP.bval"))
+    hcp_df = _find_shells(hcp_bval, 100)
+    assert hcp_df["shell_num"].max() == 4.0
+
+    # DSIQ5 should raise an exception
+    dsi_bval = np.loadtxt(load_data("schemes/DSIQ5.bval"))
+    with pytest.raises(Exception):
```
Comment (tsalo) [inline qsirecon/tests/test_interfaces.py:60]: I recommend including part of the error message in the check here.

File: qsirecon/tests/test_interfaces.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        _find_shells(dsi_bval, 100)
+
+    # Some other assorted test schemes that should fail
+    nonshelled_schemes = ["HASC55-1", "HASC55-2", "HASC92", "RAND57", "Q7"]
+    for scheme in nonshelled_schemes:
+        bval_file = Path(get_test_data_path()) / f"{scheme}.bval"
+        test_bvals = np.loadtxt(bval_file)
+
+        with pytest.raises(Exception):
+            _find_shells(test_bvals, 100)
```
Comment (tsalo) [inline qsirecon/tests/test_interfaces.py:70]: What about testing "lowest" and "highest" with these?

File: qsirecon/tests/test_interfaces.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    assert sel_bval.shape == (correct_n,)
+    sel_bvec = np.loadtxt(dwi_prefix + "_selected.bvec")
+    assert sel_bvec.shape == (3, correct_n)
+    sel_b = np.loadtxt(dwi_prefix + "_selected.b")
+    assert sel_b.shape[0] == correct_n
+    # There is no btable from this dataset because it was created
+    # before those were written in the outputs.
+    assert not Path(dwi_prefix + "_selected.txt").exists()
+
+    # Check some other sequences we might run into
```
Comment (tsalo) [inline qsirecon/tests/test_interfaces.py:47]: I recommend splitting the `_find_shells` checks into a separate test. You could parametrize over bval scheme files to make the test super simple.

**Review by tsalo — COMMENTED** [review]
I have a couple of ideas for the tests.

File: qsirecon/interfaces/gradients.py (hunk 5 above)
Comment (mattcieslak) [reply qsirecon/interfaces/gradients.py:152]: you mean including "highest", "lowest" and "0"? Those should be taken care of by the clustering

File: qsirecon/tests/test_interfaces.py (hunk 6 above)
Comment (tsalo) [reply qsirecon/tests/test_interfaces.py:60]: Here's an example: https://github.com/PennLINC/xcp_d/blob/55c09d5273ff8f15c503d1366061d45abed47d68/xcp_d/tests/test_cli_utils.py#L19-L20

File: qsirecon/interfaces/gradients.py (hunk 5 above)
Comment (tsalo) [reply qsirecon/interfaces/gradients.py:152]: No I mean if you put in "highest" for a spectrum-type set of bvals, shouldn't it raise an error? But this check (which is the one that raises an error) only applies if you use a number.

Comment (mattcieslak) [reply qsirecon/interfaces/gradients.py:152]: `_find_shells` has to be called first to create the `bval_df` needed by this function. Since an exception will get raised in `_find_shells` if the scheme isn't shelled, qsirecon won't make it to this function

File: qsirecon/interfaces/gradients.py [hunk 9]
```diff
@@ -18,33 +18,214 @@
...
+            ok_shells = bval_df["assigned_shell"][
+                np.logical_and(
+                    bval_df["assigned_shell"] >= requested_bval - max_distance,
+                    bval_df["assigned_shell"] <= requested_bval + max_distance,
+                )
+            ]
+            if len(ok_shells.unique()) > 1:
+                raise Exception(
+                    f"Unable to unambiguously select b={requested_bval}."
+                    f"instead select from {bval_df.assigned_shells.unique().tolist()}"
```
Comment (tsalo) [inline qsirecon/interfaces/gradients.py:149]:
```suggestion
                    f"Unable to unambiguously select b={requested_bval}. "
                    f"Instead select from {bval_df.assigned_shells.unique().tolist()}"
```

File: qsirecon/interfaces/gradients.py (hunk 5 above)
Comment (tsalo) [reply qsirecon/interfaces/gradients.py:152]: Ohh okay. Thanks for explaining!
